ravenml/data/write_dataset.py

```diff
@@ -0,0 +1,309 @@
+        self.obj_list = []
+    
+    def load_image_ids(self):
+        """Method goes through imagesets and is expected to populate the 'tags_df'
+            dataframe with image_ids and tags related to each image_id, as well
+            as 'image_ids' with a list of image_ids.
+
+        Args:
+        """
+        raise NotImplementedError
+
+    def interactive_filter(self):
+        """Method assumes that 'image_ids' and 'tags_df' have already been found and allows
+            user to filter through them for subsets they choose to use based on tags
+
+        Args:
+        """
+        raise NotImplementedError
+
+    def construct_all(self):
+        """Method should create objects from the image_ids given with whatever
+            information is needed for the write_dataset method to use
+
+        Args:
+        Returns:
+            object list of whatever data will be written in the dataset
+        """
+        raise NotImplementedError
+
+    @cli_spinner_wrapper("Writing out dataset locally...")
+    def write_dataset(self):
+        """Main driver, writes dataset based on objects passed from construct_all
+
+        Args:
+        """
+        raise NotImplementedError
+
+    @cli_spinner_wrapper("Writing out metadata locally...")
+    def write_metadata(self):
+        """Writes out a metadata file
+
+        Args:
+        """
+        raise NotImplementedError
+    
+    @cli_spinner_wrapper("Writing out additional files...")
+    def write_additional_files(self):
+        """Writes out additional files
+
+        Args:
+        """
+        raise NotImplementedError
+
+class DefaultDatasetWriter(DatasetWriter):
+    """Default Interface for creating datasets, methods are in order of what is expected to be 
+        called by the plugins. Plugin is expected to overload 'construct_all', 'export_data',
+        and 'write_additional_files', if the plugin expects to call 'write_dataset'. 
+
+    Methods (not in DatasetWriter):
+        write_data (objects, path, split_type): method that is overloaded by the plugin, is called
+            on by 'write_out_complete_set' in this implementation to write the contents of the objects
+            created by 'construct_all'
+        write_out_test_set (path (Path), data (list)): helper method for this implementation of
+            'write_dataset', writes out test set   
+        write_out_complete_set (path (Path), data (list)): helper method for this implementation of
+            'write_dataset', creates test and train groups and corresponding paths for plugin to write to        
+    """
+
+    def __init__(self, create: CreateInput):
+        """Method calls DatasetWriter's initialization to get all variables it needs
+
+        Args:
+            create (CreateInput): what is passed to the plugin,
+                containing configuration information
+        """
+        super().__init__(create)
+
+    def write_data(self, objects, path, split_type, *args, **kwargs):
```

Comment:
It would be nice if this method name was a bit more specific, it's kind of easy to get it mixed up with `write_dataset`. Ideally, it would indicate the kind of data it's writing, similar to `write_out_test_set`.

The problem is that our naming convention is pretty terrible, with the `test.record` and top-level holdout `test` directory being completely different things. At some point, we should really rename the `test.record` data to something like 'validation' or 'development' which are both commonly used. We also have to keep in mind that this method will eventually be used to write out the splits of k-fold cross-validation, so the naming convention has to make sense in that context as well.

However, this is a whole other discussion and out of the scope of this PR. For now, I'm leaning towards `write_out_train_split`, since even though one of the splits is called "test" it's still kind of part of the training process, and at least this clearly distinguishes the TFRecord splits from the true holdout test set. The docstring should also be updated to clarify what exactly a "train split" is and counteract the confusing-ness of one of the splits still being called "test".

Reply:
I agree on this